File: 2021/9/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n")
  .map((a) => a.split("").map(Number));

const directions = [
  [-1, 0],
  [1, 0],
  [0, -1],
  [0, 1],
];
const rMax = input.length - 1;
const cMax = input[0].length - 1;

let part1 = 0;
const candidates = [];

for (let r = 0; r < input.length; r++) {
  for (let c = 0; c < input[0].length; c++) {
    let isLowPoint = true;
    for (const [dr, dc] of directions) {
      const r2 = r + dr;
      const c2 = c + dc;
      if (r2 < 0 || r2 > rMax || c2 < 0 || c2 > cMax) continue;
      if (input[r2][c2] <= input[r][c]) {
        isLowPoint = false;
        break;
      }
    }
    if (isLowPoint) {
      part1 += input[r][c] + 1;
      candidates.push([r, c]);
    }
  }
}

console.log(part1);

function mapBasin(row, col) {
  const queue = [[row, col]];
  const visited = new Set([`${row},${col}`]);
  let size = 0;
  for (let i = 0; i < queue.length; i++) {
    size++;
    const [r, c] = queue[i];
    for (const [dr, dc] of directions) {
      const r2 = r + dr;
      const c2 = c + dc;
      if (r2 < 0 || r2 > rMax || c2 < 0 || c2 > cMax) continue;
      if (input[r2][c2] === 9) continue;
      if (visited.has(`${r2},${c2}`)) continue;
      visited.add(`${r2},${c2}`);
      queue.push([r2, c2]);
    }
  }
  return size;
}

const basinSizes = [];
for (const [r, c] of candidates) {
  basinSizes.push(mapBasin(r, c));
}

basinSizes.sort((a, b) => b - a);

console.log(basinSizes.slice(0, 3).reduce((a, b) => a * b, 1));

File: 2021/9/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	area := parseInput(strings.Split(string(data), "\n"))
	fmt.Println(solve(area))
}

var directions = [][2]int{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}

func parseInput(data []string) [][]int {
	area := make([][]int, len(data))
	for i, line := range data {
		nums := make([]int, len(line))
		for j, s := range line {
			n := s - '0'
			nums[j] = int(n)
		}
		area[i] = nums
	}
	return area
}

func calcBasinSize(area [][]int, lowPoint [2]int) int {
	visited := make(map[[2]int]struct{})
	visited[lowPoint] = struct{}{}
	var queue [][2]int
	var pos, pos2 [2]int
	queue = append(queue, lowPoint)
	for len(queue) > 0 {
		pos = queue[0]
		queue = queue[1:]
		for _, d := range directions {
			pos2[0], pos2[1] = pos[0]+d[0], pos[1]+d[1]
			if pos2[0] < 0 || pos2[1] < 0 || pos2[0] >= len(area) || pos2[1] >= len(area[0]) {
				continue
			}
			if area[pos2[0]][pos2[1]] == 9 {
				continue
			}
			if _, ok := visited[pos2]; ok {
				continue
			}
			visited[pos2] = struct{}{}
			queue = append(queue, pos2)
		}
	}
	return len(visited)
}

func updateTop3(top3 *[3]int, n int) {
	for i := range top3 {
		if n > top3[i] {
			copy(top3[i+1:], top3[i:])
			top3[i] = n
			break
		}
	}
}

func solve(area [][]int) (int, int) {
	var isLowPoint bool
	var r2, c2, part1 int
	var basinSizesTop3 [3]int
	for r, row := range area {
		for c, ht := range row {
			isLowPoint = true
			for _, d := range directions {
				r2, c2 = r+d[0], c+d[1]
				if r2 < 0 || c2 < 0 || r2 >= len(area) || c2 >= len(row) {
					continue
				}
				if area[r2][c2] <= ht {
					isLowPoint = false
					break
				}
			}
			if isLowPoint {
				part1 += ht + 1
				updateTop3(&basinSizesTop3, calcBasinSize(area, [2]int{r, c}))
			}
		}
	}
	part2 := 1
	for _, size := range basinSizesTop3 {
		part2 *= size
	}
	return part1, part2
}
